[FSDP] Check module.training for _root_cast_forward_inputs

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104223

We might erroneously cast forward inputs for the root if it doesn't
manage any handles (FSDP parameters). As a fix, pass in the module and check
its training attribute to ensure we don't cast inputs in eval mode.

Differential Revision: [D47041673](https://our.internmc.facebook.com/intern/diff/D47041673/)

rohan-varma committed:
[FSDP] Check module.training for _root_cast_forward_inputs

We might erroneously cast forward inputs for the root if it doesn't
manage any handles (FSDP parameters). As a fix, pass in the module and check
its training attribute to ensure we don't cast inputs in eval mode.

Differential Revision: [D47041673](https://our.internmc.facebook.com/intern/diff/D47041673/)

[ghstack-poisoned]

fegin reviewed (approved):
LGTM. The linter issue is real.

rohan-varma committed:
Update on "[FSDP] Check module.training for _root_cast_forward_inputs"


We might erroneously cast forward inputs for the root if it doesn't
manage any handles (FSDP parameters). As a fix, pass in the module and check
its training attribute to ensure we don't cast inputs in eval mode.

Differential Revision: [D47041673](https://our.internmc.facebook.com/intern/diff/D47041673/)

[ghstack-poisoned]

rohan-varma commented:
@pytorchbot merge -f "CI done"
